### DISCUSSION TITLE
V7 Does useController support valueAsNumber?

### DISCUSSION THREAD

**Discussion by writes-of-spring** [react-hook-form/react-hook-form#4542] [Q&A] [answered]
Does useController support value as number?

The docs for use controller say rules is an object in the same format for register, does that apply to valueAsNumber, valueAsDate and setValueAs? I know they're controlled inputs, and I can do that myself, but my use case is a useController wrapper around MUI's textfield, and the docs (and intellisense) seem to suggest I can be lazy. 

CSB example (changing ageRegister stays as number, changing ageController coerces to string)
https://codesandbox.io/s/loving-worker-t0ikv?file=/src/App.tsx

Thanks yet again!

**Comment by bluebill1049** [comment, marked as answer]
no we don't support `valueAs` in Controller, apply this instead https://react-hook-form.com/advanced-usage#TransformandParse

**Comment by shannonmoeller** [comment]
This change is super weird and inconsistent. You can use `setValueAs` with `useController` in v6 and not supporting this pattern in v7 makes `useController` inconsistent with `register`. Given that both patterns provide the `onChange` handler, why can't `rules` be the same for both cases?

**Reply by bluebill1049** [reply]
Because for the controlled component you are already controlling the value...

**Reply by shannonmoeller** [reply]
That's an invalid assumption. We prefer to use useController over register for most form fields because it lets us set `defaultValue` and gives us easy access to all of the information we need about a field including the field state and errors regardless of whether the form control is a native html element or wrapped up in a custom component. This makes it very easy to keep all field handling consistent across our forms using a single pattern with a single hook.

```jsx
export function YearField() {
  const { field, fieldState } = useController({
    name: 'year',
    defaultValue: 2021,
    rules: {
      setValueAs: (v) => Number(v) || 0, // what we're missing from v6
      required: 'Year is required',
    },
  });
  const { ref, ...restField } = field;
  const { error, invalid } = fieldState; // thank you for adding error to this object

  return (
    <Label label="Year" caption="Example: 2020" error={error}>
      <Input {...restField} error={invalid} inputRef={ref} />
    </Label>
  );
}
```

Doing this using `useFormContext`, `get(formState.errors)`, `register()`, and a custom change handler is incredibly tedious, especially when useController is 99% of the way there.

**Comment by timkindberg** [comment]
I'd like to bump this thread and see if @bluebill1049 might offer any additional insight.

We've started using this library here at Vndly (Workday) and I noticed the APIs from register and useControlled were not balanced. At first I didn't question it, but I'm starting to wonder if they could (and should) be balanced. 

The line of thinking goes even deeper though as @shannonmoeller points out... like they are not even using register at all, even in situation where it might be "preferred" if you are looking at the documentation. This begs the question(s)?

Is there a downside with using `useController` for ALL FIELDS ALL THE TIME? Is it less performant than register? It seems to provide so much extra sugar compared to register (e.g. defaultValue, fieldState, formState, etc).

**Reply by timkindberg** [reply]
FWIW I feel like I found a way to have `defaultValue` work with `register`, I asked about it here: https://github.com/react-hook-form/react-hook-form/discussions/8392

**Reply by timkindberg** [reply]
> Is there a downside with using useController for ALL FIELDS ALL THE TIME? Is it less performant than register? It seems to provide so much extra sugar compared to register (e.g. defaultValue, fieldState, formState, etc).

@shannonmoeller could you tell me if you've had any negative side effects from using useController for all fields regardless of if they could register or not?

Like I wonder what the performance impact is vs. trying to use register when possible? 🤔 Because I agree 100% that useController has a more full featured API.

**Reply by Moshyfawn** [reply]
An input used with the `Controller` API has "direct" access to it's field value, state and the form state. While the later is a `Proxy` object _(won't cause unnecessary re-renders with non-subscribed states)_, both field and it's state value are reactive properties which are going to cause re-renders for your input to read the values. That's the "trade-off".

That said, it sounds like you're comparing both `register` and `Controller` as if they're the same, when they are actually 2 different approaches to handing an input value: uncontrolled and controlled inputs. They can't have the same API a priori
